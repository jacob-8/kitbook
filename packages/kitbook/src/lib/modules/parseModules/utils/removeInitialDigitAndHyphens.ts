export function removeInitialDigitAndHyphens(string: string) {
  return string.replace(/^\d+\w?/, '')
    .replace(/-/g, ' ')
    .trim()
}

if (import.meta.vitest) {
  describe(removeInitialDigitAndHyphens, () => {
    test('replace hyphens with spaces', () => {
      expect(removeInitialDigitAndHyphens('0-Amazing-Button')).toEqual('Amazing Button')
    })
    test('removes any letters appended to digit', () => {
      expect(removeInitialDigitAndHyphens('0a-another-thing')).toEqual('another thing')
    })
  })
}
