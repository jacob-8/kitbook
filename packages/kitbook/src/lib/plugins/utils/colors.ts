export const green = '\x1B[32m'
export const red = '\x1B[31m'
export const bold = '\x1B[1m'
export const reset = '\x1B[0m'
