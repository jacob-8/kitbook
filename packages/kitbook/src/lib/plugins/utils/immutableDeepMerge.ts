/** @author inspired by [jhildenbiddle](https://stackoverflow.com/a/48218209) */
export function immutableDeepMerge(...objects: Record<string, any>[]) {
  const initialObject = {}
  const isObject = (item: any) => item && typeof item === 'object'

  return objects.reduce((previousObj, currentObj) => {
    if (currentObj === undefined)
      return previousObj

    Object.keys(currentObj).forEach((key) => {
      const previousValue = previousObj[key]
      const currentValue = currentObj[key]

      // Note that an Array is also an Object so check for array first
      if (Array.isArray(previousValue) && Array.isArray(currentValue))
        previousObj[key] = [...new Set([...previousValue, ...currentValue])]
        // if you don't want to remove duplicates, use this instead
        // previousObj[key] = [...previousValue, ...currentValue]

      // Svelte Preprocessors can be an array of objects, or just one object
      else if (isObject(previousValue) && Array.isArray(currentValue))
        previousObj[key] = [previousValue, ...currentValue]

      else if (Array.isArray(previousValue) && isObject(currentValue))
        previousObj[key] = [...previousValue, currentValue]

      else if (isObject(previousValue) && isObject(currentValue))
        previousObj[key] = immutableDeepMerge(previousValue, currentValue)
      else
        previousObj[key] = currentValue
    })

    return previousObj
  }, initialObject)
}
