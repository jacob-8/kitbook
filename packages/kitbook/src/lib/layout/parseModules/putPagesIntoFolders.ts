import type { Folder, GroupedPageMap } from '../../kitbook-types'
import { removeInitialDigitAndHyphens } from './utils/removeInitialDigitAndHyphens'

export function putPagesIntoFolders(groupedPages: GroupedPageMap): Folder {
  if (!groupedPages) {
    return {
      name: 'No pages found',
      url: '/',
      depth: 0,
    }
  }
  const pagesToOrganize = Object.values(groupedPages)

  const rootFolder: Folder = {
    name: '.',
    url: '/',
    depth: 0,
    folders: [],
    pages: [],
  }

  if (!pagesToOrganize?.length)
    return rootFolder

  pagesToOrganize.forEach((page) => {
    const pageIsJustAComposition = page.extensions.length === 1 && page.extensions[0] === 'composition'
    if (pageIsJustAComposition)
      return

    let currentFolder = rootFolder

    const path = page.url.split('/')
    path.forEach((folderName, index) => {
      const cleanedFolderName = removeInitialDigitAndHyphens(folderName)

      if (index === path.length - 1) {
        currentFolder.pages.push(page) // is a page
      }
      else if (folderName === '') {
        currentFolder = rootFolder
      }
      else {
        let folder = currentFolder.folders.find(({ name }) => name === cleanedFolderName)
        if (!folder) {
          folder = {
            name: cleanedFolderName,
            url: path
              .slice(0, index + 1)
              .join('/')
              .replace(/^\./, ''),
            depth: index,
            folders: [],
            pages: [],
          }
          currentFolder.folders.push(folder)
        }
        currentFolder = folder
      }
    })
  })
  return rootFolder
}
