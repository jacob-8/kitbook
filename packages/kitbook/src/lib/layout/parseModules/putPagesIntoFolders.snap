{
  "depth": 0,
  "folders": [
    {
      "depth": 1,
      "folders": [
        {
          "depth": 2,
          "folders": [],
          "name": "my notes",
          "pages": [
            {
              "extensions": [
                "md",
              ],
              "loadMarkdown": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "name": "unocss",
              "path": "/src/docs/my-notes/0-unocss.md",
              "url": "/docs/my-notes/0-unocss",
            },
            {
              "extensions": [
                "md",
              ],
              "loadMarkdown": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "name": "deploy to vercel",
              "path": "/src/docs/my-notes/1-deploy-to-vercel.md",
              "url": "/docs/my-notes/1-deploy-to-vercel",
            },
          ],
          "url": "/docs/my-notes",
        },
      ],
      "name": "docs",
      "pages": [
        {
          "extensions": [
            "md",
          ],
          "loadMarkdown": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "why kitbook",
          "path": "/src/docs/0-why-kitbook.md",
          "url": "/docs/0-why-kitbook",
        },
        {
          "extensions": [
            "md",
          ],
          "loadMarkdown": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "get started",
          "path": "/src/docs/1-get-started.md",
          "url": "/docs/1-get-started",
        },
        {
          "extensions": [
            "md",
          ],
          "loadMarkdown": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "you can use letters to adjust ordering",
          "path": "/src/docs/1a-you-can-use-letters-to-adjust-ordering.md",
          "url": "/docs/1a-you-can-use-letters-to-adjust-ordering",
        },
      ],
      "url": "/docs",
    },
    {
      "depth": 1,
      "folders": [
        {
          "depth": 2,
          "folders": [],
          "name": "a",
          "pages": [
            {
              "extensions": [
                "svelte",
                "variants.ts",
              ],
              "loadComponent": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "loadVariants": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "name": "C",
              "path": "/src/lib/a/C.svelte",
              "url": "/lib/a/C",
            },
            {
              "extensions": [
                "svelte",
                "md",
                "variants.ts",
                "composition",
                "first.composition",
                "second.composition",
              ],
              "loadComponent": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "loadCompositions": {
                "default": {
                  "loadModule": [Function],
                  "loadRaw": [Function],
                },
                "first": {
                  "loadModule": [Function],
                  "loadRaw": [Function],
                },
                "second": {
                  "loadModule": [Function],
                  "loadRaw": [Function],
                },
              },
              "loadMarkdown": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "loadVariants": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "name": "D",
              "path": "/src/lib/a/D.svelte",
              "url": "/lib/a/D",
            },
          ],
          "url": "/lib/a",
        },
      ],
      "name": "lib",
      "pages": [
        {
          "extensions": [
            "svelte",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "A",
          "path": "/src/lib/A.svelte",
          "url": "/lib/A",
        },
        {
          "extensions": [
            "svelte",
            "md",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "loadMarkdown": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "B",
          "path": "/src/lib/B.svelte",
          "url": "/lib/B",
        },
        {
          "extensions": [
            "md",
          ],
          "loadMarkdown": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "E",
          "path": "/src/lib/E.md",
          "url": "/lib/E",
        },
      ],
      "url": "/lib",
    },
    {
      "depth": 1,
      "folders": [
        {
          "depth": 2,
          "folders": [],
          "name": "a",
          "pages": [
            {
              "extensions": [
                "svelte",
                "md",
              ],
              "loadComponent": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "loadMarkdown": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "name": "+page",
              "path": "/src/routes/a/+page.svelte",
              "url": "/routes/a/+page",
            },
            {
              "extensions": [
                "svelte",
                "md",
              ],
              "loadComponent": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "loadMarkdown": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "name": "+layout",
              "path": "/src/routes/a/+layout.svelte",
              "url": "/routes/a/+layout",
            },
          ],
          "url": "/routes/a",
        },
        {
          "depth": 2,
          "folders": [],
          "name": "b",
          "pages": [
            {
              "extensions": [
                "svelte",
                "variants.ts",
              ],
              "loadComponent": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "loadVariants": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "name": "+page",
              "path": "/src/routes/b/+page.svelte",
              "url": "/routes/b/+page",
            },
            {
              "extensions": [
                "svelte",
                "variants.ts",
              ],
              "loadComponent": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "loadVariants": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "name": "+layout",
              "path": "/src/routes/b/+layout.svelte",
              "url": "/routes/b/+layout",
            },
          ],
          "url": "/routes/b",
        },
        {
          "depth": 2,
          "folders": [],
          "name": "c",
          "pages": [
            {
              "extensions": [
                "svelte",
                "md",
                "variants.ts",
              ],
              "loadComponent": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "loadMarkdown": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "loadVariants": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "name": "+page",
              "path": "/src/routes/c/+page.svelte",
              "url": "/routes/c/+page",
            },
            {
              "extensions": [
                "svelte",
                "md",
                "variants.ts",
              ],
              "loadComponent": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "loadMarkdown": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "loadVariants": {
                "loadModule": [Function],
                "loadRaw": [Function],
              },
              "name": "+layout",
              "path": "/src/routes/c/+layout.svelte",
              "url": "/routes/c/+layout",
            },
          ],
          "url": "/routes/c",
        },
      ],
      "name": "routes",
      "pages": [
        {
          "extensions": [
            "svelte",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "+page",
          "path": "/src/routes/+page.svelte",
          "url": "/routes/+page",
        },
        {
          "extensions": [
            "svelte",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "+layout",
          "path": "/src/routes/+layout.svelte",
          "url": "/routes/+layout",
        },
      ],
      "url": "/routes",
    },
  ],
  "name": ".",
  "pages": [
    {
      "extensions": [
        "md",
      ],
      "loadMarkdown": {
        "loadModule": [Function],
        "loadRaw": [Function],
      },
      "name": "README",
      "path": "/README.md",
      "url": "/README",
    },
    {
      "extensions": [
        "md",
      ],
      "loadMarkdown": {
        "loadModule": [Function],
        "loadRaw": [Function],
      },
      "name": "index",
      "path": "/src/index.md",
      "url": "/index",
    },
  ],
  "url": "/",
}